File: apple/HDHROpenKit/Sources/HDHROpenKit/Playback/StreamURLBuilder.swift
import Foundation

public enum StreamURLBuilder {
    public static func liveStreamURL(baseURL: URL, channelNumber: String) -> URL? {
        URL(string: "/api/streaming/stream/\(channelNumber)", relativeTo: baseURL)?.absoluteURL
    }

    public static func recordingStreamURL(
        baseURL: URL,
        playUrl: String,
        recordingId: String? = nil,
        startOffset: Double? = nil,
        audioIndex: Int? = nil
    ) -> URL? {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "url", value: playUrl)
        ]
        if let recId = recordingId {
            queryItems.append(URLQueryItem(name: "recording_id", value: recId))
        }
        if let start = startOffset, start > 0 {
            queryItems.append(URLQueryItem(name: "start", value: String(format: "%.1f", start)))
        }
        if let audio = audioIndex {
            queryItems.append(URLQueryItem(name: "audio_index", value: String(audio)))
        }

        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: true)
        components?.path = "/api/dvr/recording-stream"
        components?.queryItems = queryItems
        return components?.url
    }

    public static func hlsPlaylistURL(baseURL: URL, sessionId: String) -> URL? {
        URL(string: "/api/hls/\(sessionId)/playlist.m3u8", relativeTo: baseURL)?.absoluteURL
    }

    public static func thumbnailSpriteURL(baseURL: URL, recordingId: String, playUrl: String, recordEnd: Double? = nil) -> URL? {
        var queryItems = [URLQueryItem(name: "url", value: playUrl)]
        if let end = recordEnd {
            queryItems.append(URLQueryItem(name: "record_end", value: String(format: "%.1f", end)))
        }
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: true)
        components?.path = "/api/dvr/recording-thumbnails/\(recordingId).jpg"
        components?.queryItems = queryItems
        return components?.url
    }

    public static func thumbnailVttURL(baseURL: URL, recordingId: String, playUrl: String, recordEnd: Double? = nil) -> URL? {
        var queryItems = [URLQueryItem(name: "url", value: playUrl)]
        if let end = recordEnd {
            queryItems.append(URLQueryItem(name: "record_end", value: String(format: "%.1f", end)))
        }
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: true)
        components?.path = "/api/dvr/recording-thumbnails/\(recordingId).vtt"
        components?.queryItems = queryItems
        return components?.url
    }

    public static func captionsURL(baseURL: URL, recordingId: String, playUrl: String, recordEnd: Double? = nil) -> URL? {
        var queryItems = [
            URLQueryItem(name: "url", value: playUrl),
            URLQueryItem(name: "recording_id", value: recordingId)
        ]
        if let end = recordEnd {
            queryItems.append(URLQueryItem(name: "record_end", value: String(format: "%.1f", end)))
        }
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: true)
        components?.path = "/api/dvr/recording-captions.vtt"
        components?.queryItems = queryItems
        return components?.url
    }
}

File: android/core/src/main/kotlin/org/hdhropen/kit/playback/StreamURLBuilder.kt
package org.hdhropen.kit.playback

import java.net.URLEncoder

object StreamURLBuilder {
    fun liveStreamURL(baseURL: String, channelNumber: String, direct: Boolean = false): String {
        val base = baseURL.trimEnd('/')
        val query = if (direct) "?direct=true" else ""
        return "$base/api/streaming/stream/$channelNumber$query"
    }

    fun recordingStreamURL(
        baseURL: String,
        playUrl: String,
        recordingId: String? = null,
        startOffset: Double? = null,
        audioIndex: Int? = null,
        provider: String? = null
    ): String {
        val base = baseURL.trimEnd('/')
        val queryItems = mutableListOf<String>()
        queryItems.add("url=${URLEncoder.encode(playUrl, "UTF-8")}")
        if (recordingId != null) {
            queryItems.add("recording_id=$recordingId")
        }
        if (startOffset != null && startOffset > 0) {
            queryItems.add("start=${String.format(java.util.Locale.US, "%.1f", startOffset)}")
        }
        if (audioIndex != null) {
            queryItems.add("audio_index=$audioIndex")
        }
        if (provider != null) {
            queryItems.add("provider=$provider")
        }
        return "$base/api/dvr/recording-stream?${queryItems.joinToString("&")}"
    }

    fun hlsPlaylistURL(baseURL: String, sessionId: String): String {
        val base = baseURL.trimEnd('/')
        return "$base/api/hls/$sessionId/playlist.m3u8"
    }

    fun thumbnailSpriteURL(
        baseURL: String,
        recordingId: String,
        playUrl: String,
        recordEnd: Double? = null,
        provider: String? = null
    ): String {
        val base = baseURL.trimEnd('/')
        val query = mutableListOf<String>()
        query.add("url=${URLEncoder.encode(playUrl, "UTF-8")}")
        if (recordEnd != null) {
            query.add("record_end=${String.format(java.util.Locale.US, "%.1f", recordEnd)}")
        }
        if (provider != null) {
            query.add("provider=$provider")
        }
        return "$base/api/dvr/recording-thumbnails/$recordingId.jpg?${query.joinToString("&")}"
    }

    fun thumbnailVttURL(
        baseURL: String,
        recordingId: String,
        playUrl: String,
        recordEnd: Double? = null,
        provider: String? = null
    ): String {
        val base = baseURL.trimEnd('/')
        val query = mutableListOf<String>()
        query.add("url=${URLEncoder.encode(playUrl, "UTF-8")}")
        if (recordEnd != null) {
            query.add("record_end=${String.format(java.util.Locale.US, "%.1f", recordEnd)}")
        }
        if (provider != null) {
            query.add("provider=$provider")
        }
        return "$base/api/dvr/recording-thumbnails/$recordingId.vtt?${query.joinToString("&")}"
    }

    fun captionsURL(
        baseURL: String,
        recordingId: String,
        playUrl: String,
        recordEnd: Double? = null,
        provider: String? = null
    ): String {
        val base = baseURL.trimEnd('/')
        val query = mutableListOf<String>()
        query.add("url=${URLEncoder.encode(playUrl, "UTF-8")}")
        query.add("recording_id=$recordingId")
        if (recordEnd != null) {
            query.add("record_end=${String.format(java.util.Locale.US, "%.1f", recordEnd)}")
        }
        if (provider != null) {
            query.add("provider=$provider")
        }
        return "$base/api/dvr/recording-captions.vtt?${query.joinToString("&")}"
    }
}
